transaction/core/src/membership_proofs/mod.rs

```diff
@@ -50,6 +50,141 @@ fn hash_nil() -> [u8; 32] {
     hasher.result().try_into().unwrap()
 }
 
+/// Compose two adjacent TxOutMembershipElements into a larger TxOutMembershipElement.
+/// Fails if they are not actually adjacent.
+///
+/// This function can be used to validate a Merkle proof from the bottom up
+/// in a side-channel resistant way. The elements lowest in the tree should be combined,
+/// then the result combined with the next highest element etc. finally getting a range
+/// and hash at the top. This is then checked against the expected hash of the root element.
+///
+/// This function MUST not branch except on erroring code paths
+///
+/// Precondition: a and b are well-formed ranges, a.from <= a.to, b.from <= b.to
+/// It is the callers job to check that.
+///
+/// Arguments:
+/// * a: TxOutMembershipElement to be combined
+/// * b: TxOutMembershipElement to be combined with a, and which should be its left or right sibling
+///
+/// Returns:
+/// * A TxOutMembershipElement for the parent of a and b, with the implied hash value.
+/// * An error if a and b were not adjacent, which implies that the merkle proof was badly structured.
+///   This error can be mapped to e.g. Error::UnexpectedMembershipElement and the caller can provide context.
+pub fn compose_adjacent_membership_elements(
+    a: &TxOutMembershipElement,
+    b: &TxOutMembershipElement,
+) -> Result<TxOutMembershipElement, ()> {
+    // a is to the left if a.to matches b.from
+    // a is to the right if a.from matches b.to
+    // Because these are inclusive [,] ranges and not half-open ranges [,)
+    // we have to add one when checking if they match.
+    // wrapping_add is used to avoid creating a branch in the generated assembly
+    let a_is_left = (a.range.to.wrapping_add(1)).ct_eq(&b.range.from);
+    let b_is_left = (b.range.to.wrapping_add(1)).ct_eq(&a.range.from);
+    // If neither is to the left according to above test, then the merkle proof isn't structured
+    // correctly and we can early return with an "unexpected merkle proof" element of some kind.
+    // The client can easily ensure that the merkle proof is well-structured.
+    // If both are to the left according to this test, then one of the ranges must be reversed,
+    // contrary to the precondition.
+    if !bool::from(a_is_left ^ b_is_left) {
+        return Err(());
+    }
+
+    // Initialize the result as if a is the left and b is right,
+    // then use conditional_assign to fix it if not.
+    // The point of this is to get conditional logic without branching,
+    // subtle::conditional_assign is assumed to be side-channel resistant
```

Comment:
Nit - but would maybe reword to indicate we're not just assuming, but subtle's docs [claim to be constant time](https://docs.rs/subtle/0.3.0/subtle/trait.ConditionallyAssignable.html#tymethod.conditional_assign), e.g.

```suggestion
    // subtle::conditional_assign is implemented to be side-channel resistant
```

Reply:
that's true -- they also say this: https://github.com/dalek-cryptography/subtle#warning

my take is, code comments here should not make assertions about the correctness of code in another library, because that code can change. code comments should instead state what assumptions this code makes for its own correctness.

so its like a maintenance note, rather than like, marketing.

the marketing pitch is certainly like, subtle is made by world class cryptographers and probably not broken, and i dont want to detract from that.

i just fundamentally don't like to write code comments in one software component that makes assertions about the implementation of a completely different software component, i'd rather just document APIs, contracts, assumptions

Reply:
Got it - that makes sense to me!